### PR TITLE
Fairseq v0.10.2 compatible - generate.py updated

### PR DESCRIPTION
**Pull request by JulianneKnott**
Updated generate.py to be compatible with Fairseq v0.10.2.
Speed tbd.

### CODE REVIEW

**Comment by JiushengChen** [comment]
Avoid checking in those txt files and old backup file - "generate_old.py".

File: benchmarks/models/fs_wmt.sh [hunk 1]
```diff
@@ -9,18 +9,18 @@
 source utils.sh
 
 # MODEL - wmt16
-./benchmark.sh \
-    fairseq \
-    wmt16.en.de.32k \
-    wmt16_en_de_bpe32k/bin \
-    valid \
-    256
+# ./benchmark.sh \
```
Comment (feihugis) [inline benchmarks/models/fs_wmt.sh:12]: This is the baseline. I guess you commented them out for debugging?

File: fastseq/optimizer/fairseq/generate_old.py [hunk 2]
```diff
@@ -0,0 +1,418 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+"""Optimize fairseq-generate (v0.9.0)"""
```
Comment (feihugis) [inline fastseq/optimizer/fairseq/generate_old.py:4]: This file will not be needed?

**Review by feihugis — COMMENTED** [review]
Thanks @JulianneKnott for this PR! I did a pass and have one question for the files under `results`: are they used for comparing the outputs from fastseq and fairseq? It will be very helpful if we can add a unit test to automatically check it. We can also integrate it with our CI in the next step.

**Comment by JulianneKnott** [comment]
> Thanks @JulianneKnott for this PR! I did a pass and have one question for the files under `results`: are they used for comparing the outputs from fastseq and fairseq? It will be very helpful if we can add a unit test to automatically check it. We can also integrate it with our CI in the next step.

Yes. Didn't mean to commit them and will remove them on the next commit. Will look into adding a unit test also.

File: benchmarks/models/fs_wmt.sh (hunk 1 above)
Comment (JulianneKnott) [reply benchmarks/models/fs_wmt.sh:12]: Yes. I was/am having some trouble getting the benchmarks to work properly, so that is a work in progress.

File: fastseq/optimizer/fairseq/generate_old.py (hunk 2 above)
Comment (JulianneKnott) [reply fastseq/optimizer/fairseq/generate_old.py:4]: No. It's vestigial. Will remove it on the next commit.

**Comment by JulianneKnott** [comment]
Benchmark Info:

> Util Model Task Split BatchSize Samples Tokens Bleu Rouge Loss Perplexity Runtime(seconds) Throughput(samples/s) Throughput(tokens/s)
> fairseq_v0.10.2 bart.large.cnn cnn_dm/len-1024.bin valid 32 13367 1098036 17.93 NA NA NA 3875 3.4 283.4
> fairseq_v0.10.2+fastseq_v0.0.4 bart.large.cnn cnn_dm/len-1024.bin valid 32 13367 1098036 17.93 NA NA NA 3028 4.4 362.6

**Review by feihugis — COMMENTED** [review]
The current code looks good to me. As this change will break the other parts of the main branch, could we create another branch (e.g., fairseq-0.10.2) for this work? After finishing all the tasks, it can be merged to the main branch.

**Comment by JulianneKnott** [comment]
Updated beam search optimizer.
Benchmarks:

> Util Model Task Split BatchSize Samples Tokens Bleu Rouge Loss Perplexity Runtime(seconds) Throughput(samples/s) Throughput(tokens/s)
> fairseq_v0.10.2+fastseq_v0.0.4 bart.large.cnn cnn_dm/len-1024.bin valid 256 13367 1098278 17.90 NA NA NA 718 18.6 1529.6

**Comment by JulianneKnott** [comment]
Updated el attention optimizer
Benchmarks:

> Util Model Task Split BatchSize Samples Tokens Bleu Rouge Loss Perplexity Runtime(seconds) Throughput(samples/s) Throughput(tokens/s)
> fairseq_v0.10.2+fastseq_v0.0.4 bart.large.cnn cnn_dm/len-1024.bin valid 320 13367 1098535 17.92 NA NA NA 621 21.5 1769.0

File: fastseq/optimizer/fairseq/beam_search_optimizer.py [hunk 3]
```diff
@@ -43,21 +42,24 @@ def step(self, step, lprobs, scores):
...
                 beam_size * 2,
                 lprobs.view(bsz, -1).size(1) - 1,  # -1 so we never select pad
             ),
-            out=(self.scores_buf, self.indices_buf),
         )
-        self.beams_buf = torch.floor_divide(self.indices_buf, vocab_size)
-        self.indices_buf.fmod_(vocab_size)
-        return self.scores_buf, self.indices_buf, self.beams_buf
+        scores_buf = top_prediction[0]
+        indices_buf = top_prediction[1]
```
Comment (feihugis) [inline fastseq/optimizer/fairseq/beam_search_optimizer.py:58]: minor: how about combining these lines with line-48 `scores_buf, indices_buf = torch.topk(...)`

File: fastseq/optimizer/fairseq/beam_search_optimizer.py [hunk 4]
```diff
@@ -692,224 +598,201 @@ def get_hypo():
...
             # and dimensions: [bsz, cand_size]
             cand_bbsz_idx = cand_beams.add(bbsz_offsets)
 
-            # finalize hypotheses that end in eos, except for blacklisted ones
-            # or candidates with a score of -inf
+            # finalize hypotheses that end in eos
+            # Shape of eos_mask: (batch size, beam size)
             eos_mask = cand_indices.eq(self.eos) & cand_scores.ne(-math.inf)
-            eos_mask[:, :beam_size][blacklist] = 0
+            eos_mask[:, :beam_size][cands_to_ignore] = torch.tensor(0).to(eos_mask)
```
Comment (feihugis) [inline fastseq/optimizer/fairseq/beam_search_optimizer.py:658]: Curious: what's the purpose of using the tensor type instead of integer `0`?

File: fastseq/optimizer/fairseq/beam_search_optimizer.py [hunk 5]
```diff
@@ -692,224 +598,201 @@ def get_hypo():
...
+        for sent in range(len(finalized)):
+            scores = torch.tensor(
+                [float(elem["score"].item()) for elem in finalized[sent]]
+            )
+            _, sorted_scores_indices = torch.sort(scores, descending=True)
+            finalized[sent] = [finalized[sent][ssi] for ssi in sorted_scores_indices]
+            finalized[sent] = torch.jit.annotate(
+                List[Dict[str, Tensor]], finalized[sent]
+            )
+        return finalized
```
Comment (feihugis) [inline fastseq/optimizer/fairseq/beam_search_optimizer.py:798]: add back the missing newline

File: fastseq/optimizer/fairseq/el_attention_optimizer.py [hunk 6]
```diff
@@ -457,7 +458,7 @@ def extract_features(
         if self.project_out_dim is not None:
             x = self.project_out_dim(x)
 
-        return x, {'attn': attn, 'inner_states': inner_states}
+        return x, {"attn": [attn], "inner_states": inner_states}
```
Comment (feihugis) [inline fastseq/optimizer/fairseq/el_attention_optimizer.py:461]: Curious: why using `"` instead of `'`?

File: fastseq/optimizer/fairseq/el_attention_optimizer.py [hunk 7]
```diff
@@ -1009,443 +1029,315 @@ def merge_encoder_out(encoder_out_list: List[Optional[EncoderOut]]):
...
+        for sent in range(len(finalized)):
+            scores = torch.tensor(
+                [float(elem["score"].item()) for elem in finalized[sent]]
+            )
+            _, sorted_scores_indices = torch.sort(scores, descending=True)
+            finalized[sent] = [finalized[sent][ssi] for ssi in sorted_scores_indices]
+            finalized[sent] = torch.jit.annotate(
+                List[Dict[str, Tensor]], finalized[sent]
+            )
+        return finalized
```
Comment (feihugis) [inline fastseq/optimizer/fairseq/el_attention_optimizer.py:1343]: Ditto

File: setup.py [hunk 8]
```diff
@@ -5,8 +5,8 @@
 from torch.utils.cpp_extension import BuildExtension, CUDAExtension
 
 FASTSEQ_VERSION = '0.0.4'
-MIN_FAIRSEQ_VERSION = '0.9.0'
-MAX_FAIRSEQ_VERSION = '0.9.0'
+MIN_FAIRSEQ_VERSION = '0.10.2'
+MAX_FAIRSEQ_VERSION = '0.10.2'
```
Comment (feihugis) [inline setup.py:9]: Are these changes compatible with any 0.10.x?

File: fastseq/optimizer/fairseq/el_attention_optimizer.py (hunk 6 above)
Comment (JulianneKnott) [reply fastseq/optimizer/fairseq/el_attention_optimizer.py:461]: That's how it is in [fairseq v0.10.2](https://github.com/pytorch/fairseq/blob/83e615d66905b8ca7483122a37da1a85f13f4b8e/fairseq/models/transformer.py#L829). Not sure why it changed between 0.9.0 and 0.10.2.

File: fastseq/optimizer/fairseq/beam_search_optimizer.py (hunk 4 above)
Comment (JulianneKnott) [reply fastseq/optimizer/fairseq/beam_search_optimizer.py:659]: I updated it for consistency with [fairseq v0.10.2](https://github.com/pytorch/fairseq/blob/83e615d66905b8ca7483122a37da1a85f13f4b8e/fairseq/sequence_generator.py#L389). I'm not sure why that changed between fairseq v0.9.0 and v0.10.2. As far as I can tell, replacing it with 0 produces the same output and still passes all the fairseq tests.

File: setup.py (hunk 8 above)
Comment (JulianneKnott) [reply setup.py]: Yes. updated the min fairseq version

File: fastseq/optimizer/fairseq/beam_search_optimizer.py [hunk 9]
```diff
@@ -517,172 +665,50 @@ def _generate(self,
...
+            original_batch_idxs = sample["id"]
+        else:
+            original_batch_idxs = torch.arange(0, bsz).type_as(tokens)
 
-        reorder_state = None
-        batch_idxs = None
-        for step in range(max_len + 1):  # one extra step for EOS marker
+        for step in range(max_len + 1):
             # reorder decoder internal states based on the prev choice of beams
+            # print(f'step: {step}')
```
Comment (feihugis) [inline fastseq/optimizer/fairseq/beam_search_optimizer.py:679]: remove this debugging code?

File: fastseq/optimizer/fairseq/el_attention_optimizer.py [hunk 10]
```diff
@@ -1009,443 +1152,368 @@ def merge_encoder_out(encoder_out_list: List[Optional[EncoderOut]]):
...
-        reorder_state = None
-        batch_idxs = None
         for step in range(max_len + 1):  # one extra step for EOS marker
             # reorder decoder internal states based on the prev choice of beams
             if reorder_state is not None:
                 if batch_idxs is not None:
                     # update beam indices to take into account removed sentences
-                    corr = batch_idxs - torch.arange(
-                        batch_idxs.numel()).type_as(batch_idxs)
+                    torch.cuda.nvtx.range_push('update idx for removed sents')
```
Comment (feihugis) [inline fastseq/optimizer/fairseq/el_attention_optimizer.py:1229]: Remove these debugging code. Same to other places.

File: fastseq/optimizer/fairseq/generate.py [hunk 11]
```diff
@@ -1,27 +1,52 @@
...
 from fastseq.utils.api_decorator import replace
+from fairseq.options import add_generation_args
 
 GENERATE_FINISHED = "done"
 POSTPROCESS_FINISHED = None
 
+original_add_generation_args = add_generation_args
+
+@replace(add_generation_args)
+def add_generation_args_v1(parser):
```
Comment (feihugis) [inline fastseq/optimizer/fairseq/generate.py:30]: minor: how about changing the name from `v1` to `v2`?

File: fastseq/optimizer/fairseq/generate.py [hunk 12]
```diff
@@ -181,238 +221,263 @@ def _detokenize(self, sample, hypos):
...
+            gen_timer.n,
+            gen_timer.sum,
+            sent_through,
+            tokens_through,
+        )
+    )
     return
+
+@replace(main)
+def main_v1(args):
```
Comment (feihugis) [inline fastseq/optimizer/fairseq/generate.py:465]: minor: `main_v1` to `main_v2`?

File: fastseq/optimizer/fairseq/beam_search_optimizer.py [hunk 13]
```diff
@@ -23,18 +23,24 @@
...
 #cache optimiztion without Efficient-Lossless Attention
 USE_OPTIMIZED_CACHE_ATTN = not config.USE_EL_ATTN
 
 
 @replace(BeamSearch)
-class BeamSearchV2(BeamSearch):
-
-    def step(self, step, lprobs, scores):
-        super()._init_buffers(lprobs)
+class BeamSearch(BeamSearch):
```
[the remark this answers is not in the document]
Comment (JulianneKnott) [reply fastseq/optimizer/fairseq/beam_search_optimizer.py:33]: Yes, some of the[ fairseq tests](https://github.com/pytorch/fairseq/blob/83e615d66905b8ca7483122a37da1a85f13f4b8e/tests/test_inference_dropout.py#L57) rely on the names being the same.
